[AIRFLOW-6290] Create guide for GKE operators (#8883)
mistercrunch/Airflow: docs/howto/operator/gcp/kubernetes_engine.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

> Google Kubernetes Engine Operators

Reviewer: Can you add link to this guide in the Kubernetes Pod Operator guide? 
https://airflow.readthedocs.io/en/latest/howto/operator/kubernetes.html

```
    .. note::
        If you use `Google Kubernetes Engine <https://cloud.google.com/kubernetes-engine/>`__, use
        :ref:`GKEStartPodOperator <howto/operator:GKEStartPodOperator>`__ operaator, which
        simplifies the authorization process.
```

Author: [Done here](https://github.com/apache/airflow/pull/8883/commits/3fe1780ae0f9503988c725c6b12b2bdf9b8e6bed)

Author: [Follow up commit - had to correct the format](https://github.com/apache/airflow/pull/8883/commits/ded5cca846c7f39bccb707046e31770a7578064a)

.. exampleinclude:: ../../../../airflow/providers/google/cloud/example_dags/example_kubernetes_engine.py
    :language: python
    :dedent: 4
    :start-after: [START howto_operator_gke_create_cluster]
    :end-before: [END howto_operator_gke_create_cluster]

.. _howto/operator:GKEDeleteClusterOperator:

Delete GKE cluster
""""""""""""""""""

To delete a cluster, use
:class:`~airflow.providers.google.cloud.operators.kubernetes_engine.GKEDeleteClusterOperator`.
This would also delete all the nodes allocated to the cluster.

.. exampleinclude:: ../../../../airflow/providers/google/cloud/example_dags/example_kubernetes_engine.py
    :language: python
    :dedent: 4
    :start-after: [START howto_operator_gke_delete_cluster]
    :end-before: [END howto_operator_gke_delete_cluster]

Manage workloads on a GKE cluster
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

GKE works with containerized applications, such as those created on Docker, and deploys them to run on the cluster.
These are called workloads, and when deployed on the cluster they leverage the CPU and memory resources of the cluster
to run effectively.

.. _howto/operator:GKEStartPodOperator:

> Run a Pod on a GKE cluster

Reviewer: Can you add link to this guide?
https://airflow.readthedocs.io/en/latest/howto/operator/kubernetes.html

Author: [Think this is a duplicate of this](https://github.com/apache/airflow/pull/8883#discussion_r426998501)

Specified guide as part of addressing that comment so considering this done.

""""""""""""""""""""""""""

> To run a Kubernetes Pod on a GKE cluster, use
> :class:`~airflow.providers.google.cloud.operators.kubernetes_engine.GKEStartPodOperator`.
> It is worth noting that this extends
> :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator`
> so many of the same arguments can be used. We can enable the usage of :ref:`XCom <concepts:xcom>`
> on the operator:

Reviewer: ```suggestion
You can use two operators to run pod on a  GKE cluster:
* :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator`
* :class:`~airflow.providers.google.cloud.operators.kubernetes_engine.GKEStartPodOperator`

``GKEPodOperator`` extends ``KubernetesPodOperator`` to provide authorization using a Google credentials. This way you don't have to manage the `kube_config` file, but it will be generated automatically.  All Kubernetes (except `config_file`)  parameters are also valid for GKEStartPodOperator.

For more information about starting pods on Airflow, please look at: :ref:`howto/operator:KubernetesPodOperator` guide.

We can enable the usage of :ref:`XCom <concepts:xcom>` on the operator:
```
WDYT? I would like to show a clear difference between these two operators, because it can be embarrassing. GKEPodOperator is not magic operaotr. It's only wrapper. 

This is especially important for me because this information is not in the Cloud Composer documentation, and I often redirect users to this documentation if they want to use KubernetesPodOperator
https://cloud.google.com/composer/docs/how-to/using/using-kubernetes-pod-operator
This guide is in the Cloud Composer documentation but applies to all users.

Author: [Agree - put a version of your suggestion here](https://github.com/apache/airflow/pull/8883/commits/3ae57e2fcaed630272e28ad7d480b6cad7c45c94)

> .. exampleinclude:: ../../../../airflow/providers/google/cloud/example_dags/example_kubernetes_engine.py

Reviewer: Can you write a long description about data transfer via xcom? I am missing information about the /airflow/xcom/return.json file.

Author: [Done here](https://github.com/apache/airflow/pull/8883/commits/025aa4c869d2b435156ad42c5bf24135769eb93e)

    :language: python
    :dedent: 4
    :start-after: [START howto_operator_gke_start_pod_xcom]
    :end-before: [END howto_operator_gke_start_pod_xcom]

And then use it in other operators:

.. exampleinclude:: ../../../../airflow/providers/google/cloud/example_dags/example_kubernetes_engine.py
    :language: python
    :dedent: 4
    :start-after: [START howto_operator_gke_xcom_result]
    :end-before: [END howto_operator_gke_xcom_result]

Reference
^^^^^^^^^

For further information, look at:

* `GKE API Documentation <https://cloud.google.com/kubernetes-engine/docs/reference/rest>`__
* `Product Documentation <https://cloud.google.com/kubernetes-engine/docs/>`__
* `Kubernetes Documentation <https://kubernetes.io/docs/home/>`__

> * `Kubernetes Documentation <https://kubernetes.io/docs/home/>`__

Reviewer: ```suggestion
```

Author: [Done here](https://github.com/apache/airflow/pull/8883/commits/70c5ac9c618defca473200ce34fe760316b639fb)

This would also delete all the nodes allocated to the cluster.

.. exampleinclude:: ../../../../airflow/providers/google/cloud/example_dags/example_kubernetes_engine.py
    :language: python
    :dedent: 4
    :start-after: [START howto_operator_gke_delete_cluster]
    :end-before: [END howto_operator_gke_delete_cluster]

Manage workloads on a GKE cluster
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

GKE works with containerized applications, such as those created on Docker, and deploys them to run on the cluster.
These are called workloads, and when deployed on the cluster they leverage the CPU and memory resources of the cluster
to run effectively.

.. _howto/operator:GKEStartPodOperator:

Run a Pod on a GKE cluster
""""""""""""""""""""""""""

There are two operators available in order to run a pod on a GKE cluster:

* :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator`
* :class:`~airflow.providers.google.cloud.operators.kubernetes_engine.GKEStartPodOperator`

``GKEStartPodOperator`` extends ``KubernetesPodOperator`` to provide authorization using Google Cloud credentials.

> There is no need to manage the `kube_config` file, as it will be generated automatically.

Reviewer: ```suggestion
There is no need to manage the ``kube_config`` file, as it will be generated automatically.
```
